pyiceberg/table/__init__.py

```diff
@@ -2271,3 +2317,244 @@ def commit(self) -> Snapshot:
             )
 
         return snapshot
+
+
+class UpdateSpec:
+    _table: Table
+    _schema: Schema
+    _spec: PartitionSpec
+    _name_to_field: Dict[str, PartitionField] = {}
+    _name_to_added_field: Dict[str, PartitionField] = {}
+    _transform_to_field: Dict[Tuple[int, str], PartitionField] = {}
+    _transform_to_added_field: Dict[Tuple[int, str], PartitionField] = {}
+    _renames: Dict[str, str] = {}
+    _added_time_fields: Dict[int, PartitionField] = {}
+    _case_sensitive: bool
+    _adds: List[PartitionField]
+    _deletes: Set[int]
+    _last_assigned_partition_id: int
+    _transaction: Optional[Transaction]
+    _unassigned_field_name = 'unassigned_field_name'
+
+    def __init__(self, table: Table, transaction: Optional[Transaction] = None, case_sensitive: bool = True) -> None:
+        self._table = table
+        self._schema = table.schema()
+        self._spec = table.spec()
+        self._name_to_field = {field.name: field for field in self._spec.fields}
+        self._name_to_added_field = {}
+        self._transform_to_field = {(field.source_id, repr(field.transform)): field for field in self._spec.fields}
+        self._transform_to_added_field = {}
+        self._adds = []
+        self._deletes = set()
+        if len(table.specs()) == 1:
```

Comment:
Should we compare on the UnpartitionedSpec? It is convention that it is added by default, but it could be that some other writer does not implement this. Just to be a bit more defensive.

Reply:
Missed this, yeah I think that's a good idea

Reply:
Actually it's really important that this self._last_assigned_partition-id is set to the last assigned field ID across all specs not just the current spec. I'll fix this and add a test which demonstrates the importancce

Reply:
See https://github.com/apache/iceberg-python/pull/245#discussion_r1475571056 for what I mean